### DISCUSSION TITLE
Update ent_id of span and tokens

### DISCUSSION THREAD

**Discussion by NixBiks** [explosion/spaCy#11113] [Help: Coding & Implementations] [answered]
I have a custom component where I need to update `Span.ent_id_` and `Token.ent_id_`. How is that possible?

I've tried something like

```python
for t in span:
    t.ent_id_ = "NEW_LABEL"
```

which gives me the desired results. BUT when I then do `doc.ents = [span]` then it reset the `ent_id_` to what it was before. Anyone know how to fix this?

**Comment by NixBiks** [comment, marked as answer]
I've ended up creating a new span

```python
span = Span(doc, span.start, span.end, span_id="NEW_LABEL", label=span.label_)
doc.ents = [span]
```

which seems to work as expected

**Reply by polm** [reply]
Thanks for mentioning that, and your workaround. That's definitely weird behavior, we'll think about improving it.

**Comment by adrianeboyd** [comment]
I looked into this and although it doesn't look like there's directly a bug that we can address, this is definitely confusing because some span properties are separate copies of the token-level annotation and some are a view of the underlying token annotation.

For getting or setting `Doc.ents` there is currently this mapping between `Token` and `Span` attributes:

  * `Span.label` <-> `Token.ent_type`
  * `Span.kb_id` <-> `Token.ent_kb_id`
  * `Span.id` <-> `Token.ent_id`

The `Span.id` <-> `Token.ent_id` mapping is new as of the addition of `SpanRuler` in v3.3.1.

`Span` already had the existing `Span.ent_id` property that is just a view of the token annotation rather than an attribute of the span itself, which is also why you can't set it on the span.

For the code above, instead of setting token values within the span (that aren't reflected in the `span`) or having to create a new `Span`, you can set `Span.id`:

```python
span = doc.ents[0]
span.id_ = "NEW_LABEL"
doc.ents = [span]
```

Because the `Span.id` <-> `Token.ent_id` mapping was added after v3.0.0, there's a backoff behavior that setting `doc.ents` only overrides the current `Token.ent_id` if the `Span.id` is set (not `0` or `""`). This backoff would be removed in v4.

We'd like to make this more consistent in v4, feedback is welcome in #11328.

**Reply by NixBiks** [reply]
Glad I'm not the only one who got a bit confused here. But thank you for the thorugh explaination - I see how this arised over time.

**Reply by NixBiks** [reply]
But `pyright` does not like this

```python
span.id_ = "NEW_LABEL"
```

I'm getting `Literal["NEW_LABEL"]` is incompatible with `property` - I guess `id_` shouldn't be typed as a property as it was done in #11119 ?

**Reply by adrianeboyd** [reply]
Hmm, the types are a bit inconsistent here and that type does look wrong for a property with a setter. We'll take a look!

**Reply by NixBiks** [reply]
Let me know if there is an issue I can point to in my codebase 😸
